modules/liquidity-pools/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+		SetAdditionalCollateralRatio(AccountId, LiquidityPoolId, CurrencyId, Option<Permill>),
+		/// Set enabled trades (who, pool_id, currency_id, longs, shorts)
+		SetEnabledTrades(AccountId, LiquidityPoolId, CurrencyId, Leverages, Leverages),
+	}
+);
+
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		fn deposit_event() = default;
+
+		pub fn create_pool(origin, currency_id: T::CurrencyId) -> Result {
+			let who = ensure_signed(origin)?;
+			Self::_create_pool(who, currency_id).map_err(|e| e.into())
+		}
+
+		pub fn disable_pool(origin, pool_id: T::LiquidityPoolId) -> Result {
+			let who = ensure_signed(origin)?;
+			Self::_disable_pool(who, pool_id).map_err(|e| e.into())
+		}
+
+		pub fn remove_pool(origin, pool_id: T::LiquidityPoolId, currency_id: T::CurrencyId) -> Result {
+			let who = ensure_signed(origin)?;
+			Self::_remove_pool(who, pool_id, currency_id).map_err(|e| e.into())
+		}
+
+		pub fn deposit_liquidity(origin, pool_id: T::LiquidityPoolId, currency_id: T::CurrencyId, amount: T::Balance) -> Result {
+			let who = ensure_signed(origin)?;
+			Self::_deposit_liquidity(who, pool_id, currency_id, amount).map_err(|e| e.into())
+		}
+
+		pub fn withdraw_liquidity(origin, pool_id: T::LiquidityPoolId, currency_id: T::CurrencyId, amount: T::Balance) -> Result {
+			let who = ensure_signed(origin)?;
+			Self::_withdraw_liquidity(who, pool_id, currency_id, amount).map_err(|e| e.into())
+		}
+
+		pub fn set_spread(origin, pool_id: T::LiquidityPoolId, currency_id: T::CurrencyId, ask: Permill, bid: Permill) -> Result {
+			let who = ensure_signed(origin)?;
+			Self::_set_spread(who, pool_id, currency_id, ask, bid).map_err(|e| e.into())
+		}
+
+		pub fn set_additional_collateral_ratio(origin, pool_id: T::LiquidityPoolId, currency_id: T::CurrencyId, ratio: Option<Permill>) -> Result {
+			let who = ensure_signed(origin)?;
+			Self::_set_additional_collateral_ratio(who, pool_id, currency_id, ratio).map_err(|e| e.into())
+		}
+
+		pub fn set_enabled_trades(origin, pool_id: T::LiquidityPoolId, currency_id: T::CurrencyId, longs: Leverages, shorts: Leverages) -> Result {
+			let who = ensure_signed(origin)?;
+			Self::_set_enabled_trades(who, pool_id, currency_id, longs, shorts).map_err(|e| e.into())
+		}
+	}
+}
+
+decl_error! {
+	// LiquidityPools module errors
+	pub enum Error {
+		NoPermission,
+		CannotCreateMorePool,
+		PoolNotFound,
+		CannotRemovePool,
+		DepositFailed,
+		WithdrawFailed,
+	}
+}
+
+impl<T: Trait> Module<T> {
+	pub fn is_owner(pool_id: T::LiquidityPoolId, who: &T::AccountId) -> bool {
+		match Self::owners(pool_id) {
+			Some(id) => &id == who,
+			None => false,
+		}
+	}
+}
+
+// Private methods
+impl<T: Trait> Module<T> {
+	fn _create_pool(who: T::AccountId, currency_id: T::CurrencyId) -> result::Result<(), Error> {
+		let pool_id = Self::next_pool_id();
+		// increment next pool id
+		match pool_id.checked_add(&One::one()) {
```

Comment:
you can use `ok_or` to simplify the code a bit

`let id = pool_id.checked_add(&One::one()).ok_or(Error::CannotCreateMorePool)?;`

https://github.com/laminar-protocol/flowchain/blob/c9ccf3cc4adcdb895ea40a01b6f1204dab67423e/modules/synthetic-protocol/src/lib.rs#L298